x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java

```diff
@@ -478,4 +487,27 @@ private SSLService createSSLService(Environment environment, ResourceWatcherServ
         setSslService(sslService);
         return sslService;
     }
+
+    @Override
+    public void loadExtensions(ExtensionLoader loader) {
+        List<MutableLicenseService> licenseServiceFactories = loader.loadExtensions(MutableLicenseService.class);
+        if (licenseServiceFactories.size() > 1) {
+            throw new IllegalStateException(MutableLicenseService.class + " may not have multiple implementations");
+        } else if (licenseServiceFactories.size() == 1) {
+            setLicenseService(licenseServiceFactories.get(0));
+        }
+
+        List<StatusSupplier> xPackLicenseStateInitialStatusSupplier = loader.loadExtensions(StatusSupplier.class);
+        if (xPackLicenseStateInitialStatusSupplier.size() > 1) {
+            throw new IllegalStateException(StatusSupplier.class + " may not have multiple implementations");
+        } else if (licenseServiceFactories.size() == 1) {
+            setLicenseState(
+                new XPackLicenseState(() -> getEpochMillisSupplier().getAsLong(), xPackLicenseStateInitialStatusSupplier.get(0))
+            );
+        } else {
+            setLicenseState(
+                new XPackLicenseState(() -> getEpochMillisSupplier().getAsLong(), () -> new Status(License.OperationMode.TRIAL, true, null))
+            );
```

Comment:
I think we should either handle the default case (`else` branch) for _both_ `LicenseService` and `StatusSupplier` here or outside in `createComponents`. 

Additionally, I am not sure whether we actually want a `StatusSupplier` abstraction. `LicenseService` returns a License object from which the `status` is computed. I think it's actually more of problem to have separate LicenseService and StatusSupplier because the License might not agree with the Status.

Reply:
I can't move `LicenseService` creation into `loadExtensions` since it requires parameters that only passed to `createComponents`. 
I can't move `StatusSupplier` creation into `createComponents` because that creation is too late for usage. 

>  I think it's actually more of problem to have separate LicenseService and StatusSupplier because the License might not agree with the Status.

I actually agree, but can not fix without re-desiging how the license service and xpack license state work. The design has those as divorced concepts with at best a gentleman's agreement to when the state is updated. It coupling is a function of the implementation, not the class level design. It is further complicated by that we mix static and non-static access, promoting the objects created non-statically to static which can be referenced. Overall, I am not a fan of how these classes intersect or how they are accessed, but I am not trying to redesign but I am limited by the current design. 

The way  it works is that we rely on the call order which is Object Construction -> loadExtensions -> createComponents. We need to set the xpack license state ASAP since create components is too late since other xpack plugins rely on that state before createComponents is ever called.  

> Additionally, I am not sure whether we actually want a StatusSupplier abstraction

It is required for the SPI. I need something to abstract to differentiate between implementations. More specifically, the Class could be named InitialStatusSupplier since that what is what is doing (but didn't name it such since there is no way to enforce the intial part of the name). The SPI abstraction is only used when creating the object which means that SPI has the limitation such that the alt implementation can only ever set the state it once (which is what we want and this is not a general purpose abstraction)

Reply:
As discussed in person, rather than having a `StatusSupplier` we can derive the intitial status from the license. I have updated this PR to do that in [`ab7dd74` (#94927)](https://github.com/elastic/elasticsearch/pull/94927/commits/ab7dd744ccdb8c6fc4d576b5987e49527c597ab9)